node.ks

```diff
@@ -4,121 +4,159 @@
 // manual corrections.
+if not (defined node_bestFacing)
 global node_bestFacing is 5.   // ~5  degrees error (10 degree cone)
+if not (defined node_okFacing)
 global node_okFacing   is 20.  // ~20 degrees error (40 degree cone)
 
-local sstate is sas. // Save SAS State
+local sstate is sas. // save SAS state
+local rstate is rcs. // save RCS state
 
 // quo vadis?
-global nodeNd is nextnode.
-
-// Remember fuel level when we autostage to keep us from autostaging the
-// craft to death. This assumes that  "terminal stages" have no fuel, just
-// chutes and other descend-y things.
-global nodeStageFuelInit is 0.
+if not hasNode {
+	if nodeCreator:istype("delegate") nodeCreator().
+	if not hasNode uiFatal("Node", "No node to execute").
+}
+local nn is nextnode.
 
 // keep ship pointed at node
 sas off.
-lock NodeSteerDir to utilFaceBurn(lookdirup(nodeNd:deltav, ship:up:vector)).
-lock steering to NodeSteerDir.
+lock steerDir to lookdirup(nn:deltav, positionAt(ship,time:seconds+nn:eta)-body:position).
+lock steering to steerDir.
 
 // estimate burn direction & duration
-global nodeAccel is uiAssertAccel("Node").
-global nodeFacing is lookdirup(nodeNd:deltav, ship:facing:topvector).
-global nodeDob is (nodeNd:deltav:mag / nodeAccel).
-
-uiDebug("Orient to burn").
-// If have time, wait to ship almost align with maneuver node.
-// If have little time, wait at least to ship face in general direction of node
-// This prevents backwards burns, but still allows steering via engine thrust.
-// If ship is not rotating for some reason, will proceed anyway. (Maybe only torque source is engine gimbal?)
-local orientationOk is false.
-until orientationOk {
-  wait 0. //Noticed a performance issue with crafts with many parts. This forces the loop to wait one physics tick.
-  local steerVec is utilFaceBurn(lookdirup(nodeNd:deltav, ship:up:vector)):vector.
-  if  utilIsShipFacing(steerVec,node_bestFacing,0.5) or
-      ((nodeNd:eta <= nodeDob / 2) and utilIsShipFacing(steerVec,node_okFacing,5)) or
-      ship:angularvel:mag < 0.0001 { set orientationOk to true. }
+local resetBurnTime is burnTime = 0.
+if resetBurnTime set burnTime to burnTimeForDv(nn:deltav:mag).
+local dt is burnTime/2.
+
+local warpLoop is 2.
+until false {
+	// If have time, wait to ship almost align with maneuver node.
+	// If have little time, wait at least to ship face in general direction of node
+	// This prevents backwards burns, but still allows steering via engine thrust.
+	// If ship is not rotating for some reason, will proceed anyway. (Maybe only torque source is engine gimbal?)
+	wait 0.
+	local warped to false.
+	until utilIsShipFacing(steerDir,node_bestFacing,0.5) or
+		nn:eta <= dt and utilIsShipFacing(steerDir,node_okFacing,5) or
+		ship:angularvel:mag < 0.0001 and rcs = true
+	{
+		if ship:angularvel:mag < 0.01 rcs on.
+		stagingCheck().
+		if not warped { set warped to true. physWarp(1). }
+		wait 0.
+	}
+	if warped resetWarp().
```

Comment:
I'm not positive that the problem is here, but I think it might be.

I have a script which creates a node and executes it, with the node being created roughly 15-30 seconds before it is due to be executed.  My screen is spammed with messages about auto warp being cancelled and then started again.  Let me know if you're interested in the craft file and scripts.

Reply:
I was able to reproduce the spamming with the following line:  `run node_peri(0, time:seconds+30). run node.`  I hope that helps you find the issue!

Reply:
it is in lib_warp, sometimes kOS is unable to properly execute warpTo, although it reports that warp is stabilized, so I made a loop in warpSeconds, which works, but sometimes spams.